scripts/mod_loader/modapi/skills.lua

```diff
@@ -116,37 +171,44 @@ local function overrideProjectileOrArtillery(funcName, oldFunc)
     assert(metadataType == "projectile" or metadataType == "artillery", "This function only works for projectile or artillery weapons")
 
     SkillEffect[funcName] = function(self, source, damageInstance, projectileArt, delay)
-		
-		if not isUserdataPoint(source) then
-			delay = projectileArt
-			projectileArt = damageInstance
-			damageInstance = source
-			
-			-- adding backwards compatibility for mods made before ITB 1.2
-			-- it only works if piOrigin was set before AddProjectile/AddArtillery was called
-			source = self.piOrigin ~= Point(-INT_MAX, -INT_MAX) and self.piOrigin or nil
-		end
-		
-		if type(damageInstance) ~= 'userdata' then
-			damageInstance = {}
-		end
-		
+        if not isUserdataPoint(source) then
+            delay = projectileArt
+            projectileArt = damageInstance
+            damageInstance = source
+
+            -- adding backwards compatibility for mods made before ITB 1.2
+            -- it only works if piOrigin was set before AddProjectile/AddArtillery was called
+            source = self.piOrigin ~= Point(-INT_MAX, -INT_MAX) and self.piOrigin or nil
+        end
+
+        if type(damageInstance) ~= 'userdata' then
+            damageInstance = {}
+        end
+
         local metadataTable = {}
         metadataTable.type = metadataType
         metadataTable.projectileArt = projectileArt
         metadataTable.source = source or (Pawn and Pawn:GetSpace()) or nil
         metadataTable.target = damageInstance.loc
-        addDamageListMetadata(self[damageList], metadataTable)
-		
-		if source ~= nil then
-			oldFunc(self, source, damageInstance, projectileArt, delay or PROJ_DELAY)
-		else
-			if delay ~= nil then
-				oldFunc(self, damageInstance, projectileArt, delay)
-			else
-				oldFunc(self, damageInstance, projectileArt)
-			end
-		end
+
+        -- if the damage contains a script, add metadta first
+        if damageInstance.sScript ~= "" then
+            addDamageListMetadata(self[damageList], metadataTable, false)
+        end
```

Comment:
In this case, we could possibly split the script from the original damage instance, and insert it in a seprate damage instance after the projectile/artillery. I don't think there's anything technically preventing this, the only question is, will the weapon behave the same (with delay, without delay, etc)?

Reply:
Would there be a benefit to that? We would still be changing the list size (so indexes change), and now we have a script damage that we don't know where it came from (unless the metadata references that)

Reply:
We would have a guarantee that projectile/artillery damage instances always contain metadata, and you don't have to fish them out by searching the list, which sounds nice on paper... But when contrasted with the fact that we change the list size anyway, and this being the more complex solution overall, I think the current solution is good.